plugins/agent_bridge_mcp_client.go

```diff
@@ -0,0 +1,434 @@
+		deinitMCPClient()
+		mcpConfig = MCPServers{}
+		err := loadMCPPluginConfig(r)
+		if err != nil {
+			logger.Errorf("[+] Failed to load MCP plugin config: %s", err)
+			http.Error(rw, "Error while loading MCP configuration", http.StatusInternalServerError)
+			return
+		}
+		err = initMCPClient()
+		if err != nil {
+			logger.Errorf("[+] Failed to initialize MCP servers: %s", err)
+			http.Error(rw, "Error while loading MCP sub-system", http.StatusInternalServerError)
+			return
+		}
+		rw.WriteHeader(http.StatusOK)
+		_, _ = rw.Write([]byte("MCP Server(s) Initialized"))
+		return
+	}
+
+	// POST and Content-Type: application/nlq are expected
+	if r.URL.Path != "/mcp/" || r.Method != http.MethodPost || !isNLQContentType(r.Header.Get("Content-Type")) {
+		logger.Debugf("[+] Query is not POST or Content-Type is not %s, ignoring ...", CONTENT_TYPE_NLQ)
+		rw.WriteHeader(http.StatusBadRequest)
+		_, _ = rw.Write([]byte(fmt.Sprintf("You must use POST / with Content-Type: %s", CONTENT_TYPE_NLQ)))
+		return
+	}
+
+	nlqBytes, err := io.ReadAll(r.Body)
+	if err != nil {
+		logger.Errorf("[+] Error while reading the body: %s", err)
+		http.Error(rw, INTERNAL_ERROR_MSG, http.StatusInternalServerError)
+		return
+	}
+	nlq := string(nlqBytes)
+	logger.Debugf("[+] Process query: %v", nlq)
+	response, err := processQueryWithMCP(nlq)
+	if err != nil {
+		logger.Errorf("[+] Failed to process query: %s, err=%s", nlq, err)
+		http.Error(rw, INTERNAL_ERROR_MSG, http.StatusInternalServerError)
+		return
+	}
+
+	rw.WriteHeader(http.StatusOK)
+	_, _ = rw.Write([]byte(response))
+}
+
+func callMCPTool(toolName string, args *string) (string, error) {
+	// The arguments for the function come back as a JSON string
+	funcParams := map[string]any{}
+	if args != nil {
+		err := json.Unmarshal([]byte(*args), &funcParams)
+		if err != nil {
+			logger.Errorf("[+] Failed to unmarshal function parameters: %v", err)
+			return "", err
+		}
+	}
+
+	var mcpServerWithTool *MCPServerConfig = nil
+	for _, mcpServer := range mcpConfig {
+		for _, tool := range mcpServer.Tools {
+			if tool.Name == toolName {
+				mcpServerWithTool = mcpServer
+				break
+			}
+		}
+	}
+
+	if mcpServerWithTool == nil {
+		logger.Errorf("[+] Failed to find MCP server for tool: %s", toolName)
+		return "", fmt.Errorf("Unable to find tool '%s'", toolName)
+	}
+
+	logger.Infof("[+] Calling tool: (%s) from server (%s)", toolName, mcpServerWithTool.Name)
+
+	toolRequest := mcp.CallToolRequest{}
+	toolRequest.Params.Name = toolName
+	toolRequest.Params.Arguments = funcParams
+
+	callResult, err := mcpServerWithTool.Client.CallTool(context.TODO(), toolRequest)
+	if err != nil {
+		logger.Errorf("[+] Failed to call tool: %v", err)
+		return "", err
+	}
+
+	result := ""
+	for _, content := range callResult.Content {
+		if textContent, ok := content.(mcp.TextContent); ok {
+			result = result + textContent.Text
+		} else {
+			jsonBytes, _ := json.MarshalIndent(content, "", "  ")
```

Comment:
perhaps it is a good thing to test&manage the error on the Marshall, here. We can't affirm that the response of the MCP are json format (we support test & other, and in case of other we try to marshal it)